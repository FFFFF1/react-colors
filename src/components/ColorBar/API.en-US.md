| Property       | Description           | Type             | Default       |
|------------|----------------|------------------|--------------|
| model      | picker model | enum: 'rgb.r' 'rgb.g' 'rgb.b' 'hsv.h' 'hsv.s' 'hsv.v' | hsv.h  |
| onChange   | a callback function  | function    | -           |
| pointer    | user pointer element       | string\|ReactNode | <CirclePointer /> |
| color      | from ColorPicker | array    | -        |
| direction  | component direction  | enum: 'vertical' 'horizontal' | 'horizontal' |
| style      | container style       | object     | -           |
| pointerStyle | pointer style  | object  | -         |

