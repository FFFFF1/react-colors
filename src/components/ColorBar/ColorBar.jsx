import React, { Component, PureComponent, PropTypes } from 'react'
import reactCSS from 'reactcss'
import calcEventPosition from '../../helpers/calcEventPosition'

// todo 完善v

class ColorBar extends (PureComponent || Component) {
	static defaultProps = {
		model: PropTypes.oneOf(['r','g','b','h','s','l']),
		pointer: PropTypes.node,
		rgb: PropTypes.object,
    hsv: PropTypes.object,
    onChange: PropTypes.func
	}

	static defaultProps = {
		model: 'r',
		pointer: '',
		direction: 'vertical',
    InputModel: 'rgb',
    outputModel: 'rgb.r'
	}

	getBackground () {
		let {color, outputModel} = this.props
		let background = {
			'rgb.r': `linear-gradient(to top, rgb(0,${ Math.round(color[1]) },${ Math.round(color[2]) }), 
					rgb(255, ${ Math.round(color[1]) }, ${ Math.round(color[2]) })`,
			'rgb.g': `linear-gradient(to top, rgb(${ Math.round(color[0]) },0,${ Math.round(color[2]) }), 
					rgb(${ Math.round(color[0]) },255,${ Math.round(color[2]) })`,
			'rgb.b': `linear-gradient(to top, rgb(${ Math.round(color[0]) },${ Math.round(color[1]) },0), 
					rgb(${ Math.round(color[0]) },${ Math.round(color[1]) },255)`,	
			'hsv.h': `linear-gradient(to top, #f00 0%, #ff0 17%, #0f0 33%,
          #0ff 50%, #00f 67%, #f0f 83%, #f00 100%)`,
      'hsv.s': `linear-gradient(to top,rgba(0,0,0,${1-color[2]}),rgba(0,0,0,${1-color[2]})),
          linear-gradient(to top,hsl(${ color[0] },100%,100%),hsl(${ color[0] },100%,50%))`,
      'hsv.v': `linear-gradient(to top,rgba(0,0,0,1),transparent),
          linear-gradient(to top,hsl(${ color[0] },100%,50%),hsl(${ color[0] },100%,50%))`,
		}
		return background[outputModel]
	}

  getColor (e) {
    let {color, outputModel} = this.props
    let p = calcEventPosition(e, this.refs.container)
    let newColor = {
      'rgb.r': 255 - p.topP * 255,
      'rgb.g': 255 - p.topP * 255,
      'rgb.b': 255 - p.topP * 255,
      'hsv.h': 360 - 360 * p.topP,
      'hsv.s': 1 - p.topP,
      'hsv.v': 1 - p.topP,
    }
    return newColor[outputModel]
  }

  getPosition () {
    let {color, outputModel} = this.props 
    let position = {
      'rgb.r': 1 - color[0] / 255,
      'rgb.g': 1 - color[1] / 255,
      'rgb.b': 1 - color[2] / 255,
      'hsv.h': 1 - color[0] / 360,
      'hsv.s': 1 - color[1],
      'hsv.v': 1 - color[2]
    }
    return position[outputModel]
  }

  componentWillUnmount() {
    this.unbindEventListeners()
  }

  handleChange = (e) => {
    let newColor = this.getColor(e)
    this.props.onChange(newColor, e)
  }

  handleMouseDown = (e) => {
    this.handleChange(e, true)
    window.addEventListener('mousemove', this.handleChange)
    window.addEventListener('mouseup', this.handleMouseUp)
  }

  handleMouseUp = () => {
    this.unbindEventListeners()
  }

  unbindEventListeners() {
    window.removeEventListener('mousemove', this.handleChange)
    window.removeEventListener('mouseup', this.handleMouseUp)
  }

  render() {
    let top = this.getPosition() * 100
    console.log(top)
    const styles = reactCSS({
      'default': {
        root: {
          position: 'relative',
          display: 'inline-block',
          width: 20,
          height: 256,
          background: `linear-gradient(to right, #f00 0%, #ff0 17%, #0f0 33%,
            #0ff 50%, #00f 67%, #f0f 83%, #f00 100%)`,
          borderRadius: this.props.radius,
          boxShadow: this.props.shadow,
          cursor: 'default'
        },
        container: {
          margin: '0 2px',
          position: 'relative',
          height: '100%',
        },
        pointer: {
          position: 'absolute',
          left: 0,
          top: 0,
          width: '100%',
          height: '100%',
          transform: `translateY(${ top }%)`
        },
        slider: {
          width: '4px',
          borderRadius: '1px',
          height: '8px',
          boxShadow: '0 0 2px rgba(0, 0, 0, .6)',
          background: '#fff',
          transform: 'translate(-2px, -4px)',
        },
      },
      'vertical': {
        root: {
          background: this.getBackground(),
        },
      },
      'custom': {
        ...this.props.style
      }
    }, { 
        vertical: this.props.direction === 'vertical', 
        custom: !!this.props.style
    })

    return (
      <div style={ styles.root }>
        <div
          style={ styles.container }
          ref="container"
          onMouseDown={ this.handleMouseDown }
          onTouchMove={ this.handleChange }
          onTouchStart={ this.handleChange }
        >
          <div style={ styles.pointer }>
            { this.props.pointer ? (
              <this.props.pointer { ...this.props } />
            ) : (
              <div style={ styles.slider } />
            ) }
          </div>
        </div>
      </div>
    )
  }
}

export default ColorBar