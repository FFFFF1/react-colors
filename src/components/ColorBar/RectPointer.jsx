import React from 'react'

function RectPointer () {
	const style = {
    width: '4px',
    borderRadius: '1px',
    height: '8px',
    boxShadow: '0 0 2px rgba(0, 0, 0, .6)',
    background: '#fff',
    transform: 'translateY(-4px)',
  }

	return <div style={ style } />
}

export default RectPointer