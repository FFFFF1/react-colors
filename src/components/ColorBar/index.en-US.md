ColorBar is a two-demensional picker