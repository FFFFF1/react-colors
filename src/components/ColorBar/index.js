export {default} from './ColorBar.jsx'


/* 色度组件 */

// todo: 添加propTypes和defaultProps
// todo: 客制化slider组件
// todo: 修改组件，使可用于除了色度以外的属性

