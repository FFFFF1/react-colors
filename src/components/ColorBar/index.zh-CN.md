ColorBar，条形最为常见，以色彩空间中的一个坐标为可变值，其它坐标固定不变。目前支持 rgb 和 hsv 色彩空间中的所有坐标。

