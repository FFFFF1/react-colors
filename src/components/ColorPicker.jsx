/* 
	容器组件，将颜色属性和onChange事件传递给每个子组件，并作为设定颜色和获取颜色的接口 
	颜色输入格式：任何合法的颜色格式
	颜色输出格式 {
	  hsl: { a: 1, h: 0, l: 0.5, s: 1 },
	  hex: '#ff0000',
	  rgb: { r: 255, g: 0, b: 0, a: 1 },
	  hsv: { h: 0, s: 1, v: 1, a: 1 },
	  oldHue: 0
	}
	参考https://github.com/bgrins/TinyColor
*/

import React, {PropTypes} from 'react'
import reactCSS from 'reactcss'
import chroma from 'chroma-js'

/*function colorType (props, propName, componentName) {
    if ( props[propName] && !color.isValid(props[propName])) {
     	return new Error(
       		'Invalid color prop `' + propName + '` supplied to' +
        	' `' + componentName + '`. Validation failed.'
      	);
    }
}*/

class ColorPicker extends React.Component {
	static propTypes = {
		// 监听颜色变化
		onChange: PropTypes.func,
		// 容器样式
		style: PropTypes.object,
		// 默认颜色
		// defaultColor: colorType,
		// 设置颜色
		// color: colorType,
		colorModel: PropTypes.oneOf(['rgb','hsl','hsv','...'])
	}	


	static defaultProps = {
		onChange: () => {},
		onModelChange: () => {},
		defaultColor: 'red',
		history: [],
		colorModel: 'rgb'
	}

	constructor(props) {
		super(props);

		this.chroma = chroma(props.defaultColor, props.colorModel)
		this.state = {
			// data format:[r,g,b] i.e: [0,0,255]
			color: props.defaultColor,
			model: 'rgb.r',
			activeModel: props.colorModel
		}
	}

	componentWillReceiveProps(nextProps) {
		nextProps.color && this.setState({
			color: chroma(nextProps.color, nextProps.colorModel)
		})
	}

	shouldComponentUpdate(nextProps, nextState) {
		return true
	}

	handleChange (color, model, e) {
		this.setState({color: color, activeModel: model})
		this.props.onChange(color, e)
	}

	handleModelChange (nextModel, e) {
		this.setState({model: nextModel})
		this.props.onModelChange(nextModel, e)
	}

	getChildren (children) {
		// 为子组件传入新属性
		let {activeModel, color} = this.state
		this.chroma.set(activeModel, this.state.color)
		
		return React.Children.map(children, child => {
			let compName = child.type.name

			if(child.props['data-color']) {
				let divChildren = child.props.children 

				return React.cloneElement(child, {
					children: this.getChildren(divChildren)
				})
			}

			if(compName === 'ColorRadio') {
				return React.cloneElement(child, {
					checked: this.state.model === child.props.model,
					onChange: (m, e) => { 
						this.handleModelChange(child.props.model, e)
					}
				})
			}

			if(['ColorPanel','ColorBar','ColorInput','ColorPallete','ColorRadio'].indexOf(compName) > -1) {
				let model = child.props.model || this.state.model
				let type = model.split('.')[0] || 'rgb'
				return React.cloneElement(child, {
					onChange: (v, e) => {
						this.handleChange(v, type, e)
						child.props.onChange && child.props.onChange(v, e)
					},
					color: type === activeModel ? this.state.color : this.chroma.get(type),
					model: model
				})
			}

			if(/*自定义组件传入所有属性color,model*/false){
				let color = {
					rgb: [],
					hsv: [],
					model: [],
				}
				return React.cloneElement(child, {
					color: color,
					onChange: (value, type) => { 
						this.handleCustomChange(value, type, e)
					}
				})
			}

			return React.cloneElement(child)
		})
	}

	render () {
		let { children, style } = this.props
		const root = Object.assign({
			position: 'relative'
		}, style)

		return (
			<div style={ root }>
				{this.getChildren(children)}
			</div>
		)
	}
}

export default ColorPicker
