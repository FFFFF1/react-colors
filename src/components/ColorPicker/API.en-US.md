| 参数       | 说明           | 类型             | 默认值       |
|------------|----------------|------------------|--------------|
| color      | color | string array | -  |
| colorModel   | color input model  | function    | -           |
| onChange    | color change callback      | node | CirclePointer |
| onModelchange  | model change callback | array    | -        |
| style      | container style       | object     | -           |
