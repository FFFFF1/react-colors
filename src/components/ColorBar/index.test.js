import React from 'react'
import ColorPicker from '../ColorPicker.jsx'
import ColorBar from './ColorBar.jsx'
import { mount } from 'enzyme'


describe('ColorBar renders correctly', () => {

	it('show color', () => {
		const wrapper = mount(<ColorBar color={[127.5,0,0]} model='rgb.r'/>)
		expect(parseInt(wrapper.childAt(0).getNode().style.top)).toBe(50)
	})

	it('mousedown set color', () => {
		const colorBar = mount(<ColorBar color={[255,0,0]} model='rgb.r' onChange={(newColor) => {
			newColor[0] = parseInt(newColor[0])
			expect(newColor).toEqual([155,0,0])
		}}/>).find(ColorBar)

		colorBar.getNode().container = {
			clientWidth: 20,
			clientHeight: 256,
			getBoundingClientRect: function () {
				return { left: 100, top: 100 }
			}
		}	
		colorBar.simulate('mouseDown',{
			pageX: 105,
			pageY: 200
		})
	})
})