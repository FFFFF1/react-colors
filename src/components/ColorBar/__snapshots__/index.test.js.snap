// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorBar correctly render correctly 1`] = `
<div
  style="position:relative;display:inline-block;width:20px;height:256px;background:linear-gradient(to top, rgb(0,0,0), 
					rgb(255, 0, 0);cursor:default;"
>
  <div
    style="position:absolute;left:0;top:0%;"
  >
    <div
      style="width:4px;border-radius:1px;height:8px;box-shadow:0 0 2px rgba(0, 0, 0, .6);background:#fff;transform:translateY(-4px);"
    />
  </div>
</div>
`;

exports[`PsPointer renders correctly render correctly 1`] = `
<div
  style="position:absolute;width:40px;transform:translate(-10px, -5px);"
>
  <svg
    fill="#fff"
    height="10px"
    stroke="#000"
    style="float:left;transform:rotate(180deg);"
    version="1.1"
    width="10px"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M0 5 L6 1 L8 1 A2 2 0 0 1 9 2 L9 8 A2 2 0 0 1 8 9 L6 9 L0 5"
    />
  </svg>
  <svg
    fill="#fff"
    height="10px"
    stroke="#000"
    style="float:right;"
    version="1.1"
    width="10px"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M0 5 L6 1 L8 1 A2 2 0 0 1 9 2 L9 8 A2 2 0 0 1 8 9 L6 9 L0 5"
    />
  </svg>
</div>
`;
